utils/src/server/middleware.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! TODO
+
+/// TODO
+pub trait Middleware: Default + Send + Sync + Clone + 'static {
+	/// Intended to carry timestamp of a request, for example `std::time::Instant`. How the middleware
+	/// measures time, if at all, is entirely up to the implementation.
+	type Instant: Send + Copy;
```

Comment:
I wonder whether `Instant` is the right name; I'd rather assume very little about how it will be used and call it something like `OnRequest` perhaps. In the example, for instance, it's used as a counter, which seems like a perfectly reasonable use case too :)

I also wonder about allowing `Clone` rather than `Copy`; would this have any performance implications if the type is copyable anyway? It would perhaps open up some additional use cases. (on the other hand, we can relax restrictions later, and there aren't many use cases for an `on_request` that knows nothing about the request anyway, so I don't mind either way really!)

Reply:
Naming this one is hard, like really hard. Maybe it should be `X` to communicate its "whatever"-nature... (only half-joking). 
Or we could take the opposite stance and remove it and force `std::time::Instant` – after all we're not committing to this being the final shape of middleware; all we need is enough to cover substrate's needs.

Reply:
How about `Track`? `type Track = std::time::Instant` reads alright; `type Track = usize` is ok-ish. It probably violates all kinds of API guidelines, but this one is a bit special. :/